File: fetch-gallery.py
import urllib.request
import json
import re
import ssl

def fetch_url(url):
    ctx = ssl.create_default_context()
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    with urllib.request.urlopen(req, context=ctx) as response:
        return response.read().decode('utf-8')

def run():
    print("Fetching Lightroom album...")
    try:
        html = fetch_url('https://adobe.ly/4duIYGg')
    except Exception as e:
        print("Error fetching adobe.ly:", e)
        return

    # Extract IDs directly using Regex because SharesConfig is a JS object (not strict JSON)
    space_id_match = re.search(r'spaceAttributes:\s*\{[^}]*"id":"([^"]+)"', html)
    album_id_match = re.search(r'albumAttributes:\s*\{[^}]*"id":"([^"]+)"', html)
    base_url_match = re.search(r'"base":"([^"]+)"', html)
    
    if not (space_id_match and album_id_match and base_url_match):
        print("Could not find spaceId or albumId in HTML")
        return

    space_id = space_id_match.group(1)
    album_id = album_id_match.group(1)
    base_url = base_url_match.group(1) # https://photos.adobe.io/v2/
    
    assets_url = f"{base_url}spaces/{space_id}/albums/{album_id}/assets?embed=asset&subtype=image;video"
    print(f"Fetching assets from {assets_url}")
    
    try:
        assets_data = fetch_url(assets_url)
    except Exception as e:
        print("Error fetching assets API:", e)
        return

    assets_data = assets_data.replace('while (1) {}', '').strip()
    assets_json = json.loads(assets_data)
    
    photos = []
    for res in assets_json.get('resources', []):
        asset = res.get('asset', {})
        links = asset.get('links', {})
        if '/rels/rendition_type/2048' in links:
            rel_href = links['/rels/rendition_type/2048']['href']
            full_url = f"{base_url}spaces/{space_id}/{rel_href}?api_key=LightroomMobileWeb1"
            
            # Try to extract caption/date
            caption = "Untitled"
            try:
                payload = asset.get('payload', {})
                import_source = payload.get('importSource', {})
                date = payload.get('captureDate') or import_source.get('importTimestamp') or ""
                date_str = date.split('T')[0] if date else "Date unknown"
                filename = import_source.get('fileName') or "Photo"
                caption = f"{date_str} - {filename}"
            except Exception:
                pass
            
            photos.append({'url': full_url, 'caption': caption})
    
    print(f"Found {len(photos)} photos.")
    if not photos:
        print("No photos found, exiting.")
        return

    try:
        with open('lens.html', 'r', encoding='utf-8') as f:
            lens_html = f.read()
    except Exception as e:
        print("Error reading lens.html:", e)
        return

    # Generate new thumbnails HTML
    thumbnails_html = ''
    for idx, p in enumerate(photos):
        active_class = ' active' if idx == 0 else ''
        thumbnails_html += f'\n                <img src="{p["url"]}" data-large="{p["url"]}" class="thumbnail{active_class}" data-caption="{p["caption"]}" alt="Thumb {idx + 1}">'
    thumbnails_html += '\n            '
    
    # Replace the thumbnail track content
    lens_html = re.sub(
        r'(<div class="thumbnail-track" id="thumbnail-track">)[\s\S]*?(</div>)',
        f'\\1{thumbnails_html}\\2',
        lens_html
    )
    
    # Replace the main photo
    lens_html = re.sub(
        r'<img id="main-photo" src="[^"]+" alt="Selected Photo">',
        f'<img id="main-photo" src="{photos[0]["url"]}" alt="Selected Photo">',
        lens_html
    )
    
    # Replace the main caption
    lens_html = re.sub(
        r'(<div class="photo-caption" id="main-caption">)[\s\S]*?(</div>)',
        f'\\1\n            {photos[0]["caption"]}\n        \\2',
        lens_html
    )
    
    with open('lens.html', 'w', encoding='utf-8') as f:
        f.write(lens_html)
        
    print("lens.html updated successfully!")

if __name__ == '__main__':
    run()

File: fetch-gallery.js
const fs = require('fs');
const https = require('https');

function fetchUrl(url) {
    return new Promise((resolve, reject) => {
        https.get(url, (res) => {
            // Follow redirects
            if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
                return resolve(fetchUrl(res.headers.location));
            }
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => resolve(data));
        }).on('error', reject);
    });
}

async function run() {
    console.log("Fetching Lightroom album...");
    const html = await fetchUrl('https://adobe.ly/4duIYGg');
    
    const sharesMatch = html.match(/window\.SharesConfig = (\{.*?\});/);
    if (!sharesMatch) throw new Error("Could not find SharesConfig in HTML");
    
    const sharesConfig = JSON.parse(sharesMatch[1]);
    const spaceId = sharesConfig.spaceAttributes.id;
    const albumId = sharesConfig.albumAttributes.id;
    const baseUrl = sharesConfig.spaceAttributes.base; // https://photos.adobe.io/v2/
    
    const assetsUrl = `${baseUrl}spaces/${spaceId}/albums/${albumId}/assets?embed=asset&subtype=image;video`;
    console.log(`Fetching assets from ${assetsUrl}`);
    
    let assetsData = await fetchUrl(assetsUrl);
    assetsData = assetsData.replace('while (1) {}', '').trim();
    const assetsJson = JSON.parse(assetsData);
    
    const photos = [];
    assetsJson.resources.forEach(res => {
        if (res.asset && res.asset.links && res.asset.links['/rels/rendition_type/2048']) {
            const relHref = res.asset.links['/rels/rendition_type/2048'].href;
            const fullUrl = `${baseUrl}spaces/${spaceId}/${relHref}`;
            
            // Try to extract a caption/date
            let caption = "Untitled";
            try {
                const payload = res.asset.payload || {};
                const importSource = payload.importSource || {};
                const date = payload.captureDate || importSource.importTimestamp || "";
                const dateStr = date ? date.split('T')[0] : "Date unknown";
                const filename = importSource.fileName || "Photo";
                caption = `${dateStr} - ${filename}`;
            } catch (e) {}
            
            photos.push({ url: fullUrl, caption });
        }
    });
    
    console.log(`Found ${photos.length} photos.`);
    
    if (photos.length === 0) {
        console.log("No photos found, exiting.");
        return;
    }
    
    // Read lens.html
    let lensHtml = fs.readFileSync('lens.html', 'utf8');
    
    // Generate new thumbnails HTML
    let thumbnailsHtml = '';
    photos.forEach((p, idx) => {
        const activeClass = idx === 0 ? ' active' : '';
        thumbnailsHtml += `\n                <img src="${p.url}" data-large="${p.url}" class="thumbnail${activeClass}" data-caption="${p.caption}" alt="Thumb ${idx + 1}">`;
    });
    thumbnailsHtml += '\n            ';
    
    // Replace the thumbnail track content
    lensHtml = lensHtml.replace(
        /(<div class="thumbnail-track" id="thumbnail-track">)[\s\S]*?(<\/div>)/,
        `$1${thumbnailsHtml}$2`
    );
    
    // Replace the main photo
    lensHtml = lensHtml.replace(
        /<img id="main-photo" src="[^"]+" alt="Selected Photo">/,
        `<img id="main-photo" src="${photos[0].url}" alt="Selected Photo">`
    );
    
    // Replace the main caption
    lensHtml = lensHtml.replace(
        /(<div class="photo-caption" id="main-caption">)[\s\S]*?(<\/div>)/,
        `$1\n            ${photos[0].caption}\n        $2`
    );
    
    fs.writeFileSync('lens.html', lensHtml);
    console.log("lens.html updated successfully!");
}

run().catch(console.error);
